### DISCUSSION TITLE
Autocompletion for sentences with validation error

### DISCUSSION THREAD

**Discussion by durianwaffle** [eclipse-langium/langium#1623] [Q&A] [answered]
I am wondering if I have grammar error at other lines, how to show the autocompletion as usual for current line?
Now it will break down and show the default autocompletion items in completionFor.
And why the autocompletion items after typing ( is different from typing ctrl+space within ()?

Background.
- I implemented my own getCompletion in completion provider, and handled autocompletion for triggerChar === '(' there by passing an **empty completion list** and only show function block instances.
- I used **completionFor** for default autocompletion which will handle ctrl+space within ().

Question:
- CompletionFor has properties: `context: CompletionContext, next: NextFeature, acceptor: CompletionAcceptor` while getCompletion has properties: `document: LangiumDocument<AstNode>, params: CompletionParams` has no accepter. How to show the same autocompletion lists using the two functions? Now if I type ctrl+space in '()' I will see `RangeError: Maximum call stack size exceeded` error.
- How to show auto completion list as usual for current line if there is grammar mistakes in other lines?
- How to use markdown in the documentation of completion items? Can I highlight different parameters when I type at different positions in function ()?  e.g. ` item.documentation = 'the **documentation** of this function block...'` doesn't work. How to make it work?

**Comment by msujew** [comment, marked as answer]
Hey @durianwaffle,

> How to show the same autocompletion lists using the two functions?

I'm not entirely sure what you mean by that. `getCompletion` is the entry point into the completion algorithm - since we only get the parameters from the language client, it's all that we have to perform the completion. `completionFor` is way deeper into the algorithm, at a point where we've already identified what our next grammar element will be - both of these methods serve entirely different purposes.

> Now if I type ctrl+space in '()' I will see RangeError: Maximum call stack size exceeded error.

Does that happen when using the default completion provider or in your own implementation?

> How to show auto completion list as usual for current line if there is grammar mistakes in other lines?

Assuming the error recovery works as expected, syntax errors in other lines of the file should not influence completion. However, if the error is unrecoverable, that might lead to issues. You can probably override the [`findFeaturesAt`](https://github.com/eclipse-langium/langium/blob/b2c40e3a8f06becfa7b18b04cfbef7b7931f164a/packages/langium/src/lsp/completion/completion-provider.ts#L196-L222) method to get a better start position, but that is pretty non-trivial (which is why it isn't implemented that way yet in Langium).

> How to use markdown in the documentation of completion items?

In order to use markdown in the documentation, you need to use the `MarkupContent` interface. See our implementation in the completion provider as a reference:

https://github.com/eclipse-langium/langium/blob/b2c40e3a8f06becfa7b18b04cfbef7b7931f164a/packages/langium/src/lsp/completion/completion-provider.ts#L478-L487

Note that in general, completion is probably the most complicated (from a code/logic perspective) feature in Langium. It's incredibly hard to get it right for all situtations.

**Reply by durianwaffle** [reply]
Hi @msujew, thanks for your detailed response! It's very helpful! Here is some extra information about my question above
> How to show the same autocompletion lists using the two functions?

Yes, `getCompletion` is the entry point into the completion algorithm - since we only get the parameters from the language client, it's all that we have to perform the completion. But I called `super.completionFor(context, next, acceptor)` at the bottom. And I found that the completionForCrossReference will cause `RangeError: Maximum call stack size exceeded error issue`.

> Now if I type ctrl+space in '()' I will see RangeError: Maximum call stack size exceeded error.

It happens when I used the default implementation `completionForCrossReference` and `completionFor(context, next, acceptor)` which calls `completionForCrossReference`.

> How to show auto completion list as usual for current line if there is grammar mistakes in other lines?

Thanks for providing the link of override the [findFeaturesAt](https://github.com/eclipse-langium/langium/blob/b2c40e3a8f06becfa7b18b04cfbef7b7931f164a/packages/langium/src/lsp/completion/completion-provider.ts#L196-L222) method to get a better start position. And I am thinking maybe I can check by line number.

> How to use markdown for completionItem.documentation?

Thanks for providing the direction! I saw an example on it:

A MarkupContent literal represents a string value which content is interpreted based on its kind flag. Currently the protocol supports plaintext and markdown as markup kinds.

If the kind is markdown then the value can contain fenced code blocks like in [GitHub issues](https://help.github.com/articles/creating-and-highlighting-code-blocks/#syntax-highlighting)

Here is an example how such a string can be constructed using JavaScript / TypeScript:
```typescript
let markdown: MarkdownContent = {
 kind: MarkupKind.Markdown,
 value: [
 '# Header',
 'Some text',
 '```typescript',
 'someCode();',
 '```'
 ].join('\n')
};
```
Please Note that clients might sanitize the return markdown. A client could decide to remove HTML from the markdown to avoid script execution.

**Reply by msujew** [reply]
> And I found that the completionForCrossReference will cause RangeError: Maximum call stack size exceeded error issue.

I see. Since `completionForCrossReference` will run the scope provider (in order to find out which elements are visible at a certain position), could it be that your scope provider actually returns the stack overflow error? All we do is rethrow the error that the scope provider gives us:

https://github.com/eclipse-langium/langium/blob/b2c40e3a8f06becfa7b18b04cfbef7b7931f164a/packages/langium/src/lsp/completion/completion-provider.ts#L434-L440

**Reply by durianwaffle** [reply]
Thank you @msujew !
I have another question.
I find that **`textEdit` in completionItem can only edit the text after my current cursor.**
What can I do if I **want to edit the text before my current cursor?**
Do you have example on this?

**Reply by msujew** [reply]
@durianwaffle You can use the `CompletionItem#additionalTextEdits` property to perform any edits on the documents in addition to the completion edit.

**Reply by durianwaffle** [reply]
Thank you @msujew !

**Reply by durianwaffle** [reply]
Do you know how can I put the cursor at the end after`CompletionItemadditionalTextEdits` is done?

**Reply by msujew** [reply]
No, the completion API (on the LSP level) does not offer a way to specifiy where the cursor is supposed to be placed. IIRC it will generally be placed after the end of the `textEdit`.

**Reply by cdietrich** [reply]
depending on the usecase you might play arround with completions of kind snippet and `$0` as final cursor destination

**Reply by durianwaffle** [reply]
Thank you @cdietrich! Do you know which attribute I can use to change the cursor to the last word?
For example, I inserted something inside of a function like `MAX()` into `MAX(XYZY)` then deleted `MAX(` and also `)` using `additionalTextEdit` using replacing text by ''. But the cursor stays between XYZ and Y like `XYZ|Y` because it stays at the last position of `textEdit`

**Reply by durianwaffle** [reply]
And I am wondering if I have a grammar with 3 keywords:
1. `A`
2. `AB`
3. `B`

Then it causes false validation error under keyword `AB` because langium is not greedy and it processed keyword `AB` as two separate keywords `A` and `B` but there is not keyword_a then keyword_b grammar in my langium grammar.

How to fix this issue?

**Reply by cdietrich** [reply]
i dont think you can set it within additional edits

**Reply by durianwaffle** [reply]
I see. Do you have any idea where can I set it?

**Reply by cdietrich** [reply]
in the snippet itself

```
acceptor(context, {
      label: "my snippet",
      kind: CompletionItemKind.Snippet,
      insertText: "some keyword  $1 {\n\t$0\n}", // $1 will be the first cursor, $0 the final
      insertTextFormat: InsertTextFormat.Snippet,
      detail: "xxx",
      documentation: "yyyy",
      sortText: "3",
    })
```

**Reply by durianwaffle** [reply]
Thank you @cdietrich ! It helps a lot
